source/common/stats/thread_local_store.cc

```diff
@@ -194,11 +197,25 @@ void ThreadLocalStoreImpl::initializeThreading(Event::Dispatcher& main_thread_di
   tls_cache_ = ThreadLocal::TypedSlot<TlsCache>::makeUnique(tls);
   tls_cache_->set(
       [](Event::Dispatcher&) -> std::shared_ptr<TlsCache> { return std::make_shared<TlsCache>(); });
+  tls_ = tls;
 }
 
 void ThreadLocalStoreImpl::shutdownThreading() {
   // This will block both future cache fills as well as cache flushes.
   shutting_down_ = true;
+
+  if (tls_.has_value()) {
+    ASSERT(tls_->isShutdown());
+  }
```

Comment:
maybe do `ASSERT(!tls_.has_value() || tls_->isShutdown())`? Right now this turns into an empty if block in NDEBUG builds (which would get optimized out, so it doesn't matter that much)